jcl/src/java.base/share/classes/java/lang/String.java

```diff
@@ -857,47 +667,3830 @@ public String(char[] data) {
+	 */
+	public String substring(int start) {
+		if (start == 0) {
+			return this;
+		}
+
+		int len = lengthInternal();
+
+		if (0 <= start && start <= len) {
+			boolean isCompressed = false;
+
+			// Check if the String is compressed
+			if (enableCompression && (null == compressionFlag || coder == LATIN1)) {
+				isCompressed = true;
+			}
+
+			return new String(value, start, len - start, isCompressed, enableSharingInSubstringWhenOffsetIsZero);
+		} else {
+			throw new StringIndexOutOfBoundsException(start);
+		}
+	}
+
+	/**
+	 * Copies a range of characters.
+	 *
+	 * @param start
+	 *			  the offset of the first character
+	 * @param end
+	 *			  the offset one past the last character
+	 * @return a String containing the characters from start to end - 1
+	 *
+	 * @throws IndexOutOfBoundsException
+	 *				when {@code start < 0, start > end} or {@code end > length()}
+	 */
+	public String substring(int start, int end) {
+		int len = lengthInternal();
+
+		if (start == 0 && end == len) {
+			return this;
+		}
+
+		if (0 <= start && start <= end && end <= len) {
+			boolean isCompressed = false;
+			// Check if the String is compressed
+			if (enableCompression && (null == compressionFlag || coder == LATIN1)) {
+				isCompressed = true;
+			}
+
+			return new String(value, start, end - start, isCompressed, enableSharingInSubstringWhenOffsetIsZero);
+		} else {
+			throw new StringIndexOutOfBoundsException(start);
+		}
+	}
+
+	/**
+	 * Copies the characters in this String to a character array.
+	 *
+	 * @return a character array containing the characters of this String
+	 */
+	public char[] toCharArray() {
+		int len = lengthInternal();
+
+		char[] buffer = new char[len];
+
+		// Check if the String is compressed
+		if (enableCompression && (null == compressionFlag || coder == LATIN1)) {
+			decompress(value, 0, buffer, 0, len);
+		} else {
+			decompressedArrayCopy(value, 0, buffer, 0, len);
+		}
+
+		return buffer;
+	}
+
+	/**
+	 * Converts the characters in this String to lowercase, using the default Locale. To convert to lower case independent of any locale, use
+	 * toLowerCase(Locale.ROOT).
+	 *
+	 * @return a new String containing the lowercase characters equivalent to the characters in this String
+	 */
+	public String toLowerCase() {
+		return toLowerCase(Locale.getDefault());
+	}
+
+	private int toLowerCase(int codePoint) {
+		if (codePoint < 128) {
+			if ('A' <= codePoint && codePoint <= 'Z') {
+				return codePoint + ('a' - 'A');
+			} else {
+				return codePoint;
+			}
+		} else {
+			return Character.toLowerCase(codePoint);
+		}
+	}
+
+	private int toUpperCase(int codePoint) {
+		if (codePoint < 128) {
+			if ('a' <= codePoint && codePoint <= 'z') {
+				return codePoint - ('a' - 'A');
+			} else {
+				return codePoint;
+			}
+		} else {
+			return Character.toUpperCase(codePoint);
+		}
+	}
+
+// Some of the data below originated from the Unicode Character Database file
```

Comment:
This comment and the Unicode license below is not needed here since Java 9 doesn't encode the unicode data in String. This notice is in the Java 8 copy of String where the unicode data is encoded.

Reply:
Fixed in bba0fb8